{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb9afaa8",
   "metadata": {},
   "source": [
    "# How to Query OpenTargets Genetics for L2G Scores"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a3eb23",
   "metadata": {},
   "source": [
    "### 1. Create Google Cloud Credentials\n",
    "\n",
    "This step explains how to setup a Google Cloud Service Account to use Google BigQuery. Querying the first 1TB/month of public BigQuery is free, but there are fees beyond that, so you will need to link payment information to your credentials to run queries. Unless you are doing very large queries, you should not reach the 1TB limit.\n",
    "\n",
    "1. Go to https://console.cloud.google.com/\n",
    "2. If you do not have a project already, create one.\n",
    "3. Go to https://console.cloud.google.com/iam-admin/serviceaccounts & click `+ CREATE SERVICE ACCOUNT`\n",
    "4. Enter a descriptive Service Account name like \"groberts-query-opentargets\" & set the role to \"owner\"\n",
    "5. Once back on the Service Accounts main page, click the \"Actions\" menu button > \"Manage Keys\"\n",
    "6. `ADD KEY` > `Create New Key` > Select`JSON`.\n",
    "7. A `.json` file will automatically download. **Be careful with your `.json`. You will need to call this file, but DO NOT upload it to GitHub as these are your private credentials and malicious actors could use your credentials to rack up expensive queries**\n",
    "\n",
    "Once your service account is setup, the query below shows you how to access data in the BigQuery database\n",
    "`bigquery-public-data.open_targets_genetics`.  You can view the table schema in the OpenTargets Genetics database by going to \n",
    "https://console.cloud.google.com/bigquery?p=bigquery-public-data&d=open_targets_genetics&page=dataset .\n",
    "\n",
    "Note that there is a separate database for the OpenTargets Platform `bigquery-public-data.open_targets_platform` that contains additional information that you may wish to map to the OpenTargets Genetics data you are pulling in, for example known drugs that target GWAS genes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2939f42",
   "metadata": {},
   "source": [
    "## 2. Setup to Run a BigQuery\n",
    "\n",
    "Setup the BigQuery client"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "305aaefd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup credentials for BigQuery (.json generated in 7 steps above)\n",
    "\n",
    "from google.cloud import bigquery\n",
    "credentials_path = '/home/robertg1/.ssh/test-bigquery-ot-956f8a01208f.json' #Do not push this file to GitHub!\n",
    "client = bigquery.Client.from_service_account_json(credentials_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7088039c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic Imports\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "97836434",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_json_format_l2g_columns(df, col=\"trait_efos\"):\n",
    "    df[col] = df[col].apply(lambda x: x[\"list\"][0][\"element\"] if len(x[\"list\"]) == 1 else \", \".join([i[\"element\"] for i in x[\"list\"]]))\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "472f60de",
   "metadata": {},
   "source": [
    "## 3. Do an Initial Query to Identify All Available Lipid GWAS to Pull In"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "20ea9a90",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_studies = '''\n",
    "SELECT *\n",
    "FROM `bigquery-public-data.open_targets_genetics.studies`\n",
    "'''\n",
    "# Note - as defined, this query returns \n",
    "query_studies_job = client.query(query_studies)\n",
    "\n",
    "# Convert the query results to a Pandas dataframe\n",
    "studies_query_res = query_studies_job.to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a8a35c89",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get trait_efos in a simpler format for querying\n",
    "studies = process_json_format_l2g_columns(studies_query_res.copy())\n",
    "\n",
    "#query the lipid EFOs that we want\n",
    "efo_ids = [\"EFO_0004611\", \"EFO_0004612\", \"EFO_0004530\"]\n",
    "studies_of_interest = studies.query(\"trait_efos in @efo_ids\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "483f8c42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 39 unique studies in this query\n"
     ]
    }
   ],
   "source": [
    "#filter to some high-quality studies in the last 10 years\n",
    "\n",
    "min_pub_date = pd.to_datetime(\"2014-01-01\")\n",
    "studies_of_interest.loc[:, \"pub_date\"] = pd.to_datetime(studies_of_interest[\"pub_date\"])\n",
    "studies_to_pull = (\n",
    "    studies_of_interest.query(\"trait_efos in @efo_ids and num_assoc_loci >= 20 and pub_date >= @min_pub_date and not trait_reported.str.contains('\\[|drinker\\]', case=False)\", engine=\"python\")\n",
    "    .sort_values(by=\"n_initial\", ascending=False)\n",
    "    .groupby([\"pmid\", \"trait_efos\"])\n",
    "    .head(1)  # Keep only the top row in each group (optional)\n",
    "    .reset_index()\n",
    ")\n",
    "print(f\"There are {studies_to_pull.shape[0]} unique studies in this query\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8d81cd32",
   "metadata": {},
   "outputs": [],
   "source": [
    "#define a few extra and add to the list\n",
    "finngen = [\"E4_HYPERCHOL\", \"E4_HYPERLIPNAS\", \"E4_HYPERLIPMIX\"]\n",
    "ukb = [\"SAIGE_272_1\", \"SAIGE_272_11\"]\n",
    "all_gwas_to_pull = list(studies_to_pull[\"study_id\"]) + finngen + ukb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76cdd6c8",
   "metadata": {},
   "source": [
    "## 4. Use the Study IDs to Perform the Full Query\n",
    "\n",
    "Here, we will take the study IDs selected above and use those to query all genome-wide significant loci, all available co-localization data, and all available L2G scores for those studies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b8bbfb81",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>study_id_1</th>\n",
       "      <th>ancestry_initial</th>\n",
       "      <th>ancestry_replication</th>\n",
       "      <th>n_cases</th>\n",
       "      <th>n_initial</th>\n",
       "      <th>...</th>\n",
       "      <th>right_phenotype</th>\n",
       "      <th>coloc_h0</th>\n",
       "      <th>coloc_h1</th>\n",
       "      <th>coloc_h2</th>\n",
       "      <th>coloc_h3</th>\n",
       "      <th>coloc_h4</th>\n",
       "      <th>coloc_h4_h3</th>\n",
       "      <th>coloc_log2_h4_h3</th>\n",
       "      <th>pval</th>\n",
       "      <th>rn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6949</th>\n",
       "      <td>GCST002897</td>\n",
       "      <td>11</td>\n",
       "      <td>116778201</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GCST002897</td>\n",
       "      <td>{'list': [{'element': 'European=62166'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>62166</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15006</th>\n",
       "      <td>GCST002897</td>\n",
       "      <td>11</td>\n",
       "      <td>116778201</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GCST002897</td>\n",
       "      <td>{'list': [{'element': 'European=62166'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>62166</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16734</th>\n",
       "      <td>GCST002897</td>\n",
       "      <td>11</td>\n",
       "      <td>116778201</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GCST002897</td>\n",
       "      <td>{'list': [{'element': 'European=62166'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>62166</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30520</th>\n",
       "      <td>GCST002897</td>\n",
       "      <td>11</td>\n",
       "      <td>116778201</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GCST002897</td>\n",
       "      <td>{'list': [{'element': 'European=62166'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>62166</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36507</th>\n",
       "      <td>GCST002897</td>\n",
       "      <td>11</td>\n",
       "      <td>116778201</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>GCST002897</td>\n",
       "      <td>{'list': [{'element': 'European=62166'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>62166</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.000000e-157</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83500</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>11</td>\n",
       "      <td>65705287</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>{'list': [{'element': 'European=406276'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>33242</td>\n",
       "      <td>406276</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83502</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>11</td>\n",
       "      <td>65705287</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>{'list': [{'element': 'European=406276'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>33242</td>\n",
       "      <td>406276</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000213445</td>\n",
       "      <td>0.001266</td>\n",
       "      <td>0.032150</td>\n",
       "      <td>0.001497</td>\n",
       "      <td>0.037081</td>\n",
       "      <td>0.928007</td>\n",
       "      <td>25.026268</td>\n",
       "      <td>4.645371</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85291</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>11</td>\n",
       "      <td>65705287</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>{'list': [{'element': 'European=406276'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>33242</td>\n",
       "      <td>406276</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000172922</td>\n",
       "      <td>0.007134</td>\n",
       "      <td>0.181186</td>\n",
       "      <td>0.006094</td>\n",
       "      <td>0.154124</td>\n",
       "      <td>0.651462</td>\n",
       "      <td>4.226866</td>\n",
       "      <td>2.079588</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85293</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>11</td>\n",
       "      <td>65705287</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>{'list': [{'element': 'European=406276'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>33242</td>\n",
       "      <td>406276</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86138</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>11</td>\n",
       "      <td>65705287</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>{'list': [{'element': 'European=406276'}]}</td>\n",
       "      <td>{'list': []}</td>\n",
       "      <td>33242</td>\n",
       "      <td>406276</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000175573</td>\n",
       "      <td>0.030454</td>\n",
       "      <td>0.764416</td>\n",
       "      <td>0.003371</td>\n",
       "      <td>0.084500</td>\n",
       "      <td>0.117259</td>\n",
       "      <td>1.387675</td>\n",
       "      <td>0.472670</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86986 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           study_id chrom        pos ref alt    study_id_1  \\\n",
       "6949     GCST002897    11  116778201   G   C    GCST002897   \n",
       "15006    GCST002897    11  116778201   G   C    GCST002897   \n",
       "16734    GCST002897    11  116778201   G   C    GCST002897   \n",
       "30520    GCST002897    11  116778201   G   C    GCST002897   \n",
       "36507    GCST002897    11  116778201   G   C    GCST002897   \n",
       "...             ...   ...        ...  ..  ..           ...   \n",
       "83500  SAIGE_272_11    11   65705287   A   G  SAIGE_272_11   \n",
       "83502  SAIGE_272_11    11   65705287   A   G  SAIGE_272_11   \n",
       "85291  SAIGE_272_11    11   65705287   A   G  SAIGE_272_11   \n",
       "85293  SAIGE_272_11    11   65705287   A   G  SAIGE_272_11   \n",
       "86138  SAIGE_272_11    11   65705287   A   G  SAIGE_272_11   \n",
       "\n",
       "                                 ancestry_initial ancestry_replication  \\\n",
       "6949    {'list': [{'element': 'European=62166'}]}         {'list': []}   \n",
       "15006   {'list': [{'element': 'European=62166'}]}         {'list': []}   \n",
       "16734   {'list': [{'element': 'European=62166'}]}         {'list': []}   \n",
       "30520   {'list': [{'element': 'European=62166'}]}         {'list': []}   \n",
       "36507   {'list': [{'element': 'European=62166'}]}         {'list': []}   \n",
       "...                                           ...                  ...   \n",
       "83500  {'list': [{'element': 'European=406276'}]}         {'list': []}   \n",
       "83502  {'list': [{'element': 'European=406276'}]}         {'list': []}   \n",
       "85291  {'list': [{'element': 'European=406276'}]}         {'list': []}   \n",
       "85293  {'list': [{'element': 'European=406276'}]}         {'list': []}   \n",
       "86138  {'list': [{'element': 'European=406276'}]}         {'list': []}   \n",
       "\n",
       "       n_cases  n_initial  ...  right_phenotype  coloc_h0  coloc_h1  coloc_h2  \\\n",
       "6949      <NA>      62166  ...             None       NaN       NaN       NaN   \n",
       "15006     <NA>      62166  ...             None       NaN       NaN       NaN   \n",
       "16734     <NA>      62166  ...             None       NaN       NaN       NaN   \n",
       "30520     <NA>      62166  ...             None       NaN       NaN       NaN   \n",
       "36507     <NA>      62166  ...             None       NaN       NaN       NaN   \n",
       "...        ...        ...  ...              ...       ...       ...       ...   \n",
       "83500    33242     406276  ...             None       NaN       NaN       NaN   \n",
       "83502    33242     406276  ...  ENSG00000213445  0.001266  0.032150  0.001497   \n",
       "85291    33242     406276  ...  ENSG00000172922  0.007134  0.181186  0.006094   \n",
       "85293    33242     406276  ...             None       NaN       NaN       NaN   \n",
       "86138    33242     406276  ...  ENSG00000175573  0.030454  0.764416  0.003371   \n",
       "\n",
       "       coloc_h3  coloc_h4  coloc_h4_h3  coloc_log2_h4_h3           pval rn  \n",
       "6949        NaN       NaN          NaN               NaN  2.000000e-157  1  \n",
       "15006       NaN       NaN          NaN               NaN  2.000000e-157  1  \n",
       "16734       NaN       NaN          NaN               NaN  2.000000e-157  1  \n",
       "30520       NaN       NaN          NaN               NaN  2.000000e-157  1  \n",
       "36507       NaN       NaN          NaN               NaN  2.000000e-157  1  \n",
       "...         ...       ...          ...               ...            ... ..  \n",
       "83500       NaN       NaN          NaN               NaN   3.490000e-08  1  \n",
       "83502  0.037081  0.928007    25.026268          4.645371   3.490000e-08  1  \n",
       "85291  0.154124  0.651462     4.226866          2.079588   3.490000e-08  1  \n",
       "85293       NaN       NaN          NaN               NaN   3.490000e-08  1  \n",
       "86138  0.084500  0.117259     1.387675          0.472670   3.490000e-08  1  \n",
       "\n",
       "[86986 rows x 38 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Optional: Define a list of GWAS Catalog IDs/SAIGE/FINNGEN/NEALE2 GWAS IDs to keep\n",
    "#           This can be useful in selecting the highest quality/latest GWAS for each trait.\n",
    "study_ids_to_keep = all_gwas_to_pull\n",
    "\n",
    "# Construct parameterized query\n",
    "query = '''\n",
    "WITH ranked_genes AS (\n",
    "  SELECT locus2gene.study_id, \n",
    "       locus2gene.chrom, locus2gene.pos, locus2gene.ref, locus2gene.alt, \n",
    "       study_metadata.*, \n",
    "       locus2gene.gene_id, genes.gene_name, locus2gene.y_proba_full_model,\n",
    "       coloc.right_study, coloc.right_type, coloc.right_bio_feature, coloc.right_phenotype,\n",
    "       coloc.coloc_h0, coloc.coloc_h1, coloc.coloc_h2, coloc.coloc_h3, coloc.coloc_h4,\n",
    "       coloc.coloc_h4_h3, coloc.coloc_log2_h4_h3,\n",
    "       lead_variants.pval,\n",
    "       ROW_NUMBER() OVER(PARTITION BY locus2gene.study_id, genes.gene_name, coloc.right_study,\n",
    "       coloc.right_bio_feature ORDER BY lead_variants.pval) AS rn\n",
    "   \n",
    "  FROM `bigquery-public-data.open_targets_genetics.locus2gene` AS locus2gene\n",
    "   \n",
    "  -- Get lead variant P-values\n",
    "  INNER JOIN `bigquery-public-data.open_targets_genetics.variant_disease` AS lead_variants\n",
    "    ON locus2gene.pos = lead_variants.lead_pos\n",
    "      AND locus2gene.chrom = lead_variants.lead_chrom\n",
    "      AND locus2gene.study_id = lead_variants.study_id\n",
    "   \n",
    "  -- Get colocoalization stats\n",
    "  LEFT JOIN `bigquery-public-data.open_targets_genetics.variant_disease_coloc` AS coloc\n",
    "    ON (locus2gene.gene_id = coloc.right_gene_id\n",
    "      AND lead_variants.lead_chrom = coloc.left_chrom\n",
    "      AND lead_variants.lead_pos = coloc.left_pos\n",
    "      AND lead_variants.lead_ref = coloc.left_ref\n",
    "      AND lead_variants.lead_alt = coloc.left_alt\n",
    "      AND lead_variants.study_id = coloc.left_study\n",
    "      AND coloc.right_type IN ('eqtl', 'pqtl'))\n",
    "\n",
    "  -- Get GWAS metadata\n",
    "  INNER JOIN `bigquery-public-data.open_targets_genetics.studies` AS study_metadata\n",
    "    ON locus2gene.study_id = study_metadata.study_id\n",
    "   \n",
    "  -- Get HGNC IDs\n",
    "  INNER JOIN `bigquery-public-data.open_targets_genetics.genes` AS genes\n",
    "    ON locus2gene.gene_id = genes.gene_id\n",
    "   \n",
    "   -- Filter on the study IDs in the list study_ids_to_keep\n",
    "   WHERE locus2gene.study_id IN UNNEST(@study_ids_to_keep)\n",
    "   \n",
    "   -- Remove the \"raw\" Neale lab results -- I'm not sure what this is\n",
    "   AND locus2gene.study_id NOT LIKE '%raw%'\n",
    ")\n",
    "\n",
    "SELECT * FROM ranked_genes WHERE rn = 1;\n",
    "'''\n",
    "\n",
    "# Set query parameters\n",
    "job_config = bigquery.QueryJobConfig(\n",
    "    query_parameters=[\n",
    "        bigquery.ArrayQueryParameter(\"study_ids_to_keep\", \"STRING\", study_ids_to_keep)\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Run the query\n",
    "# Note - as defined, this query returns \n",
    "query_job = client.query(query, job_config=job_config)\n",
    "\n",
    "# Convert the query results to a Pandas dataframe\n",
    "l2g = query_job.to_dataframe()\n",
    "l2g.sort_values(by=['study_id', 'pval'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "75cdaf7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "l2g.to_csv('OpenTargets_L2G_Coloc_Lipids.noQC.6May24.csv.gz', compression=\"gzip\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a09808b",
   "metadata": {},
   "source": [
    "## 4. Get the Credible Sets Associated with Each Locus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23aaa605",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>tag_variant_id</th>\n",
       "      <th>postprob</th>\n",
       "      <th>is95_credset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>8:58407830:A:G</td>\n",
       "      <td>8:58420135:G:A</td>\n",
       "      <td>0.052411</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>3:136402758:A:C</td>\n",
       "      <td>3:136384425:C:T</td>\n",
       "      <td>0.017859</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>5:157051633:C:T</td>\n",
       "      <td>5:158443054:A:T</td>\n",
       "      <td>0.002977</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>8:19967156:C:T</td>\n",
       "      <td>8:18405036:T:C</td>\n",
       "      <td>0.001851</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>19:11081053:C:T</td>\n",
       "      <td>19:11079976:G:A</td>\n",
       "      <td>0.024773</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21312</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>19:11082415:T:G</td>\n",
       "      <td>19:11079398:G:A</td>\n",
       "      <td>0.049589</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21313</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>16:72067626:C:G</td>\n",
       "      <td>16:72041694:A:T</td>\n",
       "      <td>0.217111</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21314</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>19:11082415:T:G</td>\n",
       "      <td>19:11091630:G:T</td>\n",
       "      <td>0.016592</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21315</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>8:58423997:C:T</td>\n",
       "      <td>8:58418723:T:G</td>\n",
       "      <td>0.061966</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21316</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>8:58423997:C:T</td>\n",
       "      <td>8:58474778:T:C</td>\n",
       "      <td>0.023207</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21317 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           study_id  lead_variant_id   tag_variant_id  postprob  is95_credset\n",
       "0       SAIGE_272_1   8:58407830:A:G   8:58420135:G:A  0.052411          True\n",
       "1       SAIGE_272_1  3:136402758:A:C  3:136384425:C:T  0.017859          True\n",
       "2       SAIGE_272_1  5:157051633:C:T  5:158443054:A:T  0.002977          True\n",
       "3       SAIGE_272_1   8:19967156:C:T   8:18405036:T:C  0.001851          True\n",
       "4       SAIGE_272_1  19:11081053:C:T  19:11079976:G:A  0.024773          True\n",
       "...             ...              ...              ...       ...           ...\n",
       "21312  SAIGE_272_11  19:11082415:T:G  19:11079398:G:A  0.049589          True\n",
       "21313  SAIGE_272_11  16:72067626:C:G  16:72041694:A:T  0.217111          True\n",
       "21314  SAIGE_272_11  19:11082415:T:G  19:11091630:G:T  0.016592          True\n",
       "21315  SAIGE_272_11   8:58423997:C:T   8:58418723:T:G  0.061966          True\n",
       "21316  SAIGE_272_11   8:58423997:C:T   8:58474778:T:C  0.023207          True\n",
       "\n",
       "[21317 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cred_set_query = '''\n",
    "SELECT credible_sets.study_id, credible_sets.lead_variant_id, credible_sets.tag_variant_id, credible_sets.postprob, credible_sets.is95_credset\n",
    "FROM `bigquery-public-data.open_targets_genetics.variant_disease_credset` AS credible_sets\n",
    "WHERE credible_sets.study_id IN UNNEST(@study_ids_to_keep)\n",
    "AND credible_sets.is95_credset = true;\n",
    "'''\n",
    "\n",
    "# Run the query\n",
    "# Note - as defined, this query returns \n",
    "cred_set_query_job = client.query(cred_set_query, job_config=job_config)\n",
    "\n",
    "# Convert the query results to a Pandas dataframe\n",
    "cred_sets = cred_set_query_job.to_dataframe()\n",
    "cred_sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f1578bc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>tag_variant_id</th>\n",
       "      <th>postprob</th>\n",
       "      <th>is95_credset</th>\n",
       "      <th>position</th>\n",
       "      <th>cred_set_min_pos</th>\n",
       "      <th>cred_set_max_pos</th>\n",
       "      <th>cred_set_size_bp</th>\n",
       "      <th>credset_n_variants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12246</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2285</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:112161596:G:A</td>\n",
       "      <td>10:112161596:G:A</td>\n",
       "      <td>0.165411</td>\n",
       "      <td>True</td>\n",
       "      <td>112161596</td>\n",
       "      <td>112150963</td>\n",
       "      <td>112189906</td>\n",
       "      <td>38943</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7105</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:112161596:G:A</td>\n",
       "      <td>10:112180571:T:C</td>\n",
       "      <td>0.121886</td>\n",
       "      <td>True</td>\n",
       "      <td>112180571</td>\n",
       "      <td>112150963</td>\n",
       "      <td>112189906</td>\n",
       "      <td>38943</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13220</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:112161596:G:A</td>\n",
       "      <td>10:112174626:G:C</td>\n",
       "      <td>0.071861</td>\n",
       "      <td>True</td>\n",
       "      <td>112174626</td>\n",
       "      <td>112150963</td>\n",
       "      <td>112189906</td>\n",
       "      <td>38943</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1164</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:112161596:G:A</td>\n",
       "      <td>10:112161401:T:A</td>\n",
       "      <td>0.061224</td>\n",
       "      <td>True</td>\n",
       "      <td>112161401</td>\n",
       "      <td>112150963</td>\n",
       "      <td>112189906</td>\n",
       "      <td>38943</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21055</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22124451:A:G</td>\n",
       "      <td>0.006718</td>\n",
       "      <td>True</td>\n",
       "      <td>22124451</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20970</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22124631:A:G</td>\n",
       "      <td>0.006672</td>\n",
       "      <td>True</td>\n",
       "      <td>22124631</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21066</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22115960:A:G</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>True</td>\n",
       "      <td>22115960</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20574</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22124478:A:G</td>\n",
       "      <td>0.005949</td>\n",
       "      <td>True</td>\n",
       "      <td>22124478</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20665</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21317 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           study_id   lead_variant_id    tag_variant_id  postprob  \\\n",
       "12246  GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "2285   GCST90002412  10:112161596:G:A  10:112161596:G:A  0.165411   \n",
       "7105   GCST90002412  10:112161596:G:A  10:112180571:T:C  0.121886   \n",
       "13220  GCST90002412  10:112161596:G:A  10:112174626:G:C  0.071861   \n",
       "1164   GCST90002412  10:112161596:G:A  10:112161401:T:A  0.061224   \n",
       "...             ...               ...               ...       ...   \n",
       "21055  SAIGE_272_11    9:22076072:T:C    9:22124451:A:G  0.006718   \n",
       "20970  SAIGE_272_11    9:22076072:T:C    9:22124631:A:G  0.006672   \n",
       "21066  SAIGE_272_11    9:22076072:T:C    9:22115960:A:G  0.006066   \n",
       "20574  SAIGE_272_11    9:22076072:T:C    9:22124478:A:G  0.005949   \n",
       "20665  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "\n",
       "       is95_credset   position  cred_set_min_pos  cred_set_max_pos  \\\n",
       "12246          True  100315722         100315722         100315722   \n",
       "2285           True  112161596         112150963         112189906   \n",
       "7105           True  112180571         112150963         112189906   \n",
       "13220          True  112174626         112150963         112189906   \n",
       "1164           True  112161401         112150963         112189906   \n",
       "...             ...        ...               ...               ...   \n",
       "21055          True   22124451          22072265          22125504   \n",
       "20970          True   22124631          22072265          22125504   \n",
       "21066          True   22115960          22072265          22125504   \n",
       "20574          True   22124478          22072265          22125504   \n",
       "20665          True   22125504          22072265          22125504   \n",
       "\n",
       "       cred_set_size_bp  credset_n_variants  \n",
       "12246                 0                   1  \n",
       "2285              38943                  21  \n",
       "7105              38943                  21  \n",
       "13220             38943                  21  \n",
       "1164              38943                  21  \n",
       "...                 ...                 ...  \n",
       "21055             53239                  40  \n",
       "20970             53239                  40  \n",
       "21066             53239                  40  \n",
       "20574             53239                  40  \n",
       "20665             53239                  40  \n",
       "\n",
       "[21317 rows x 10 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_credset_intervals(df):\n",
    "    df['position'] = df['tag_variant_id'].str.split(':', expand=True)[1].astype(int)\n",
    "    grouped = df.groupby('lead_variant_id')\n",
    "\n",
    "    df['cred_set_min_pos'] = grouped['position'].transform('min')\n",
    "    df['cred_set_max_pos'] = grouped['position'].transform('max')\n",
    "    df['cred_set_size_bp'] = df['cred_set_max_pos'] - df['cred_set_min_pos']\n",
    "    df['credset_n_variants'] = grouped['tag_variant_id'].transform('nunique')\n",
    "    df=df.sort_values(by=[\"study_id\", \"lead_variant_id\", \"postprob\"], ascending=[True, True, False])\n",
    "    return(df)\n",
    "\n",
    "cred_sets_annotated = get_credset_intervals(cred_sets.copy())\n",
    "cred_sets_annotated"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd2c5ff",
   "metadata": {},
   "source": [
    "# 5. Add V2G Scores for Each Credible Set Variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "857ae181",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18847"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_gene_ids = l2g[\"gene_id\"].unique().tolist()\n",
    "unique_tag_variants = cred_sets_annotated[\"tag_variant_id\"].unique().tolist()\n",
    "len(unique_tag_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "14ef4e06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpred_max_label</th>\n",
       "      <th>v2g_fpred_max_score</th>\n",
       "      <th>v2g_qtl_score_q</th>\n",
       "      <th>v2g_interval_score_q</th>\n",
       "      <th>v2g_distance_score_q</th>\n",
       "      <th>distance_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16:31070704:T:C</td>\n",
       "      <td>ENSG00000099377</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.7</td>\n",
       "      <td>0.8</td>\n",
       "      <td>85497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16:31068513:C:A</td>\n",
       "      <td>ENSG00000099377</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.7</td>\n",
       "      <td>0.8</td>\n",
       "      <td>83306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:150966095:G:A</td>\n",
       "      <td>ENSG00000117360</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16:31079086:G:C</td>\n",
       "      <td>ENSG00000099377</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.8</td>\n",
       "      <td>93879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16:31057070:G:A</td>\n",
       "      <td>ENSG00000099377</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.9</td>\n",
       "      <td>71863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259823</th>\n",
       "      <td>5:123586079:A:G</td>\n",
       "      <td>ENSG00000151304</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259824</th>\n",
       "      <td>2:65052280:A:G</td>\n",
       "      <td>ENSG00000143952</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259825</th>\n",
       "      <td>1:150399540:T:G</td>\n",
       "      <td>ENSG00000143379</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259826</th>\n",
       "      <td>2:218546480:A:G</td>\n",
       "      <td>ENSG00000187736</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259827</th>\n",
       "      <td>3:129516858:G:C</td>\n",
       "      <td>ENSG00000114656</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>259828 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             variant_id          gene_id fpred_max_label  v2g_fpred_max_score  \\\n",
       "0       16:31070704:T:C  ENSG00000099377              NA                  NaN   \n",
       "1       16:31068513:C:A  ENSG00000099377              NA                  NaN   \n",
       "2       1:150966095:G:A  ENSG00000117360              NA                  NaN   \n",
       "3       16:31079086:G:C  ENSG00000099377              NA                  NaN   \n",
       "4       16:31057070:G:A  ENSG00000099377              NA                  NaN   \n",
       "...                 ...              ...             ...                  ...   \n",
       "259823  5:123586079:A:G  ENSG00000151304              NA                  NaN   \n",
       "259824   2:65052280:A:G  ENSG00000143952              NA                  NaN   \n",
       "259825  1:150399540:T:G  ENSG00000143379              NA                  NaN   \n",
       "259826  2:218546480:A:G  ENSG00000187736              NA                  NaN   \n",
       "259827  3:129516858:G:C  ENSG00000114656              NA                  NaN   \n",
       "\n",
       "        v2g_qtl_score_q  v2g_interval_score_q  v2g_distance_score_q  \\\n",
       "0                   0.9                   0.7                   0.8   \n",
       "1                   0.9                   0.7                   0.8   \n",
       "2                   0.0                   NaN                   NaN   \n",
       "3                   0.9                   0.8                   0.8   \n",
       "4                   0.9                   0.6                   0.9   \n",
       "...                 ...                   ...                   ...   \n",
       "259823              NaN                   0.4                   NaN   \n",
       "259824              NaN                   0.4                   NaN   \n",
       "259825              NaN                   0.0                   NaN   \n",
       "259826              NaN                   0.0                   NaN   \n",
       "259827              NaN                   0.0                   NaN   \n",
       "\n",
       "        distance_to_tss  \n",
       "0                 85497  \n",
       "1                 83306  \n",
       "2                  <NA>  \n",
       "3                 93879  \n",
       "4                 71863  \n",
       "...                 ...  \n",
       "259823             <NA>  \n",
       "259824             <NA>  \n",
       "259825             <NA>  \n",
       "259826             <NA>  \n",
       "259827             <NA>  \n",
       "\n",
       "[259828 rows x 8 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v2g_query = '''\n",
    "WITH v2g_concat AS (\n",
    "  SELECT *,\n",
    "         CONCAT(v2g.chr_id, ':', CAST(v2g.position AS STRING), ':', v2g.ref_allele, ':', v2g.alt_allele) AS variant_id\n",
    "  FROM `bigquery-public-data.open_targets_genetics.variant_gene` AS v2g\n",
    ")\n",
    "\n",
    "SELECT variant_id,\n",
    "       gene_id,\n",
    "       CASE WHEN COUNT(DISTINCT fpred_max_label) = 1 THEN MAX(fpred_max_label) ELSE 'NA' END AS fpred_max_label,\n",
    "       MAX(fpred_max_score) AS v2g_fpred_max_score,\n",
    "       MAX(qtl_score_q) AS v2g_qtl_score_q,\n",
    "       MAX(interval_score_q) AS v2g_interval_score_q,\n",
    "       MAX(distance_score_q) AS v2g_distance_score_q,\n",
    "       MIN(d) AS distance_to_tss\n",
    "FROM v2g_concat\n",
    "WHERE variant_id IN UNNEST (@unique_tag_variants)\n",
    "GROUP BY variant_id, gene_id;\n",
    "\n",
    "'''\n",
    "\n",
    "# Run the query\n",
    "# Note - as defined, this query returns \n",
    "# Set query parameters\n",
    "v2g_job_config = bigquery.QueryJobConfig(\n",
    "    query_parameters=[\n",
    "        bigquery.ArrayQueryParameter(\"unique_tag_variants\", \"STRING\", unique_tag_variants),\n",
    "        bigquery.ArrayQueryParameter(\"unique_gene_ids\", \"STRING\", unique_gene_ids)\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Convert the query results to a Pandas dataframe\n",
    "v2g_query_job = client.query(v2g_query, job_config=v2g_job_config)\n",
    "v2g = v2g_query_job.to_dataframe()\n",
    "v2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f1e4408f",
   "metadata": {},
   "outputs": [],
   "source": [
    "v2g.to_csv('OpenTargets_Variant2GeneScores_Lipids.6May24.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4baa6e6a",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "## 6. Perform Joins and Cleanup Dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1e191ac1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait_efos</th>\n",
       "      <th>trait_reported</th>\n",
       "      <th>source</th>\n",
       "      <th>study_id</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pub_date</th>\n",
       "      <th>n_cases</th>\n",
       "      <th>num_assoc_loci</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>lead_variant_pval</th>\n",
       "      <th>...</th>\n",
       "      <th>coloc_type</th>\n",
       "      <th>coloc_tissue</th>\n",
       "      <th>coloc_probe_transcript_id</th>\n",
       "      <th>coloc_h0</th>\n",
       "      <th>coloc_h1</th>\n",
       "      <th>coloc_h2</th>\n",
       "      <th>coloc_h3</th>\n",
       "      <th>coloc_h4</th>\n",
       "      <th>coloc_h4_h3</th>\n",
       "      <th>coloc_log2_h4_h3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32231</th>\n",
       "      <td>EFO_0003774</td>\n",
       "      <td>Hyperlipidemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>35844</td>\n",
       "      <td>45</td>\n",
       "      <td>19:44908822:C:T</td>\n",
       "      <td>2.100000e-72</td>\n",
       "      <td>...</td>\n",
       "      <td>pqtl</td>\n",
       "      <td>UBERON_0001969</td>\n",
       "      <td>APOE_5312_49_3</td>\n",
       "      <td>1.080201e-60</td>\n",
       "      <td>6.397096e-56</td>\n",
       "      <td>1.699519e-08</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>154213.068110</td>\n",
       "      <td>17.234565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86974</th>\n",
       "      <td>EFO_0003774</td>\n",
       "      <td>Hyperlipidemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>35844</td>\n",
       "      <td>45</td>\n",
       "      <td>19:44908822:C:T</td>\n",
       "      <td>2.100000e-72</td>\n",
       "      <td>...</td>\n",
       "      <td>eqtl</td>\n",
       "      <td>NEUTROPHIL_CD15</td>\n",
       "      <td>ILMN_1683475</td>\n",
       "      <td>2.650833e-53</td>\n",
       "      <td>4.170651e-01</td>\n",
       "      <td>4.167357e-54</td>\n",
       "      <td>0.065049</td>\n",
       "      <td>0.517886</td>\n",
       "      <td>7.961524</td>\n",
       "      <td>2.993045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17559</th>\n",
       "      <td>EFO_0003774</td>\n",
       "      <td>Hyperlipidemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>35844</td>\n",
       "      <td>45</td>\n",
       "      <td>19:44908822:C:T</td>\n",
       "      <td>2.100000e-72</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32230</th>\n",
       "      <td>EFO_0003774</td>\n",
       "      <td>Hyperlipidemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>35844</td>\n",
       "      <td>45</td>\n",
       "      <td>19:44908822:C:T</td>\n",
       "      <td>2.100000e-72</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45234</th>\n",
       "      <td>EFO_0003774</td>\n",
       "      <td>Hyperlipidemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_1</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>35844</td>\n",
       "      <td>45</td>\n",
       "      <td>19:44908822:C:T</td>\n",
       "      <td>2.100000e-72</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44395</th>\n",
       "      <td>HP_0003124</td>\n",
       "      <td>Hypercholesterolemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>33242</td>\n",
       "      <td>40</td>\n",
       "      <td>11:65705287:A:G</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25413</th>\n",
       "      <td>HP_0003124</td>\n",
       "      <td>Hypercholesterolemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>33242</td>\n",
       "      <td>40</td>\n",
       "      <td>11:65705287:A:G</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85293</th>\n",
       "      <td>HP_0003124</td>\n",
       "      <td>Hypercholesterolemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>33242</td>\n",
       "      <td>40</td>\n",
       "      <td>11:65705287:A:G</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76516</th>\n",
       "      <td>HP_0003124</td>\n",
       "      <td>Hypercholesterolemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>33242</td>\n",
       "      <td>40</td>\n",
       "      <td>11:65705287:A:G</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>eqtl</td>\n",
       "      <td>PLATELET</td>\n",
       "      <td>ILMN_1652777</td>\n",
       "      <td>2.531117e-02</td>\n",
       "      <td>6.438328e-01</td>\n",
       "      <td>4.154343e-03</td>\n",
       "      <td>0.105452</td>\n",
       "      <td>0.221250</td>\n",
       "      <td>2.098121</td>\n",
       "      <td>1.069098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56487</th>\n",
       "      <td>HP_0003124</td>\n",
       "      <td>Hypercholesterolemia</td>\n",
       "      <td>SAIGE</td>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>None</td>\n",
       "      <td>2018-10-24</td>\n",
       "      <td>33242</td>\n",
       "      <td>40</td>\n",
       "      <td>11:65705287:A:G</td>\n",
       "      <td>3.490000e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86986 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        trait_efos        trait_reported source      study_id  pmid  \\\n",
       "32231  EFO_0003774        Hyperlipidemia  SAIGE   SAIGE_272_1  None   \n",
       "86974  EFO_0003774        Hyperlipidemia  SAIGE   SAIGE_272_1  None   \n",
       "17559  EFO_0003774        Hyperlipidemia  SAIGE   SAIGE_272_1  None   \n",
       "32230  EFO_0003774        Hyperlipidemia  SAIGE   SAIGE_272_1  None   \n",
       "45234  EFO_0003774        Hyperlipidemia  SAIGE   SAIGE_272_1  None   \n",
       "...            ...                   ...    ...           ...   ...   \n",
       "44395   HP_0003124  Hypercholesterolemia  SAIGE  SAIGE_272_11  None   \n",
       "25413   HP_0003124  Hypercholesterolemia  SAIGE  SAIGE_272_11  None   \n",
       "85293   HP_0003124  Hypercholesterolemia  SAIGE  SAIGE_272_11  None   \n",
       "76516   HP_0003124  Hypercholesterolemia  SAIGE  SAIGE_272_11  None   \n",
       "56487   HP_0003124  Hypercholesterolemia  SAIGE  SAIGE_272_11  None   \n",
       "\n",
       "         pub_date  n_cases  num_assoc_loci  lead_variant_id  \\\n",
       "32231  2018-10-24    35844              45  19:44908822:C:T   \n",
       "86974  2018-10-24    35844              45  19:44908822:C:T   \n",
       "17559  2018-10-24    35844              45  19:44908822:C:T   \n",
       "32230  2018-10-24    35844              45  19:44908822:C:T   \n",
       "45234  2018-10-24    35844              45  19:44908822:C:T   \n",
       "...           ...      ...             ...              ...   \n",
       "44395  2018-10-24    33242              40  11:65705287:A:G   \n",
       "25413  2018-10-24    33242              40  11:65705287:A:G   \n",
       "85293  2018-10-24    33242              40  11:65705287:A:G   \n",
       "76516  2018-10-24    33242              40  11:65705287:A:G   \n",
       "56487  2018-10-24    33242              40  11:65705287:A:G   \n",
       "\n",
       "       lead_variant_pval  ... coloc_type     coloc_tissue  \\\n",
       "32231       2.100000e-72  ...       pqtl   UBERON_0001969   \n",
       "86974       2.100000e-72  ...       eqtl  NEUTROPHIL_CD15   \n",
       "17559       2.100000e-72  ...       None             None   \n",
       "32230       2.100000e-72  ...       None             None   \n",
       "45234       2.100000e-72  ...       None             None   \n",
       "...                  ...  ...        ...              ...   \n",
       "44395       3.490000e-08  ...       None             None   \n",
       "25413       3.490000e-08  ...       None             None   \n",
       "85293       3.490000e-08  ...       None             None   \n",
       "76516       3.490000e-08  ...       eqtl         PLATELET   \n",
       "56487       3.490000e-08  ...       None             None   \n",
       "\n",
       "      coloc_probe_transcript_id      coloc_h0      coloc_h1      coloc_h2  \\\n",
       "32231            APOE_5312_49_3  1.080201e-60  6.397096e-56  1.699519e-08   \n",
       "86974              ILMN_1683475  2.650833e-53  4.170651e-01  4.167357e-54   \n",
       "17559                      None           NaN           NaN           NaN   \n",
       "32230                      None           NaN           NaN           NaN   \n",
       "45234                      None           NaN           NaN           NaN   \n",
       "...                         ...           ...           ...           ...   \n",
       "44395                      None           NaN           NaN           NaN   \n",
       "25413                      None           NaN           NaN           NaN   \n",
       "85293                      None           NaN           NaN           NaN   \n",
       "76516              ILMN_1652777  2.531117e-02  6.438328e-01  4.154343e-03   \n",
       "56487                      None           NaN           NaN           NaN   \n",
       "\n",
       "       coloc_h3  coloc_h4    coloc_h4_h3  coloc_log2_h4_h3  \n",
       "32231  0.000006  0.999993  154213.068110         17.234565  \n",
       "86974  0.065049  0.517886       7.961524          2.993045  \n",
       "17559       NaN       NaN            NaN               NaN  \n",
       "32230       NaN       NaN            NaN               NaN  \n",
       "45234       NaN       NaN            NaN               NaN  \n",
       "...         ...       ...            ...               ...  \n",
       "44395       NaN       NaN            NaN               NaN  \n",
       "25413       NaN       NaN            NaN               NaN  \n",
       "85293       NaN       NaN            NaN               NaN  \n",
       "76516  0.105452  0.221250       2.098121          1.069098  \n",
       "56487       NaN       NaN            NaN               NaN  \n",
       "\n",
       "[86986 rows x 23 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_l2g = process_json_format_l2g_columns(l2g.copy())\n",
    "\n",
    "#create the variant IDs\n",
    "filtered_l2g[\"lead_variant_id\"] = filtered_l2g.chrom.astype(str) + ':' + filtered_l2g.pos.astype(str) + ':' + filtered_l2g.ref.astype(str) + ':' + filtered_l2g.alt.astype(str)\n",
    "\n",
    "#for continuous traits, fill n_cases with n_initial\n",
    "filtered_l2g[\"n_cases\"] = filtered_l2g[\"n_cases\"].fillna(filtered_l2g[\"n_initial\"])\n",
    "\n",
    "#rename some columns\n",
    "filtered_l2g.rename(columns={'y_proba_full_model': 'L2G',\n",
    "                             'pval' : 'lead_variant_pval',\n",
    "                             'right_study' : 'coloc_study',\n",
    "                             'right_type' : 'coloc_type',\n",
    "                             'right_bio_feature' : 'coloc_tissue',\n",
    "                             'right_phenotype' : 'coloc_probe_transcript_id'},\n",
    "                    inplace=True)\n",
    "\n",
    "filtered_l2g = filtered_l2g.sort_values(by=['trait_efos', 'study_id', 'lead_variant_pval', 'L2G', 'coloc_h4'],\n",
    "                                        ascending=[True, True, True, False, False])\n",
    "\n",
    "#filter down the columns to make this df a bit easier to work with\n",
    "desired_columns = ['trait_efos', 'trait_reported',\n",
    "                   'source', 'study_id', 'pmid',\n",
    "                   'pub_date', 'n_cases','num_assoc_loci',\n",
    "                   'lead_variant_id', 'lead_variant_pval',\n",
    "                   'gene_name', 'L2G',\n",
    "                   'coloc_study', 'coloc_type', 'coloc_tissue', 'coloc_probe_transcript_id',\n",
    "                   'coloc_h0', 'coloc_h1', 'coloc_h2', 'coloc_h3', 'coloc_h4',\n",
    "                  'coloc_h4_h3', 'coloc_log2_h4_h3']\n",
    "filtered_l2g = filtered_l2g[desired_columns].drop_duplicates()\n",
    "filtered_l2g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "06dc223c",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_l2g.to_csv('OpenTargets_L2G_Coloc_Lipids.Cleaned.6May24.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dd3828c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>tag_variant_id</th>\n",
       "      <th>postprob</th>\n",
       "      <th>is95_credset</th>\n",
       "      <th>position</th>\n",
       "      <th>cred_set_min_pos</th>\n",
       "      <th>cred_set_max_pos</th>\n",
       "      <th>cred_set_size_bp</th>\n",
       "      <th>credset_n_variants</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpred_max_label</th>\n",
       "      <th>v2g_fpred_max_score</th>\n",
       "      <th>v2g_qtl_score_q</th>\n",
       "      <th>v2g_interval_score_q</th>\n",
       "      <th>v2g_distance_score_q</th>\n",
       "      <th>distance_to_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000120054</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5</td>\n",
       "      <td>233853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000107566</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7</td>\n",
       "      <td>129689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000107593</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000196072</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.9</td>\n",
       "      <td>29042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000166135</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>213350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294531</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22124478:A:G</td>\n",
       "      <td>0.005949</td>\n",
       "      <td>True</td>\n",
       "      <td>22124478</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22124478:A:G</td>\n",
       "      <td>ENSG00000147883</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.8</td>\n",
       "      <td>115173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294532</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.7</td>\n",
       "      <td>130203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294533</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000147883</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.8</td>\n",
       "      <td>116199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294534</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000176399</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4</td>\n",
       "      <td>321320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294535</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000099810</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7</td>\n",
       "      <td>0.4</td>\n",
       "      <td>322868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272750 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            study_id   lead_variant_id    tag_variant_id  postprob  \\\n",
       "0       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "1       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "2       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "4       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "5       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "...              ...               ...               ...       ...   \n",
       "294531  SAIGE_272_11    9:22076072:T:C    9:22124478:A:G  0.005949   \n",
       "294532  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "294533  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "294534  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "294535  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "\n",
       "        is95_credset   position  cred_set_min_pos  cred_set_max_pos  \\\n",
       "0               True  100315722         100315722         100315722   \n",
       "1               True  100315722         100315722         100315722   \n",
       "2               True  100315722         100315722         100315722   \n",
       "4               True  100315722         100315722         100315722   \n",
       "5               True  100315722         100315722         100315722   \n",
       "...              ...        ...               ...               ...   \n",
       "294531          True   22124478          22072265          22125504   \n",
       "294532          True   22125504          22072265          22125504   \n",
       "294533          True   22125504          22072265          22125504   \n",
       "294534          True   22125504          22072265          22125504   \n",
       "294535          True   22125504          22072265          22125504   \n",
       "\n",
       "        cred_set_size_bp  credset_n_variants        variant_id  \\\n",
       "0                      0                   1  10:100315722:G:A   \n",
       "1                      0                   1  10:100315722:G:A   \n",
       "2                      0                   1  10:100315722:G:A   \n",
       "4                      0                   1  10:100315722:G:A   \n",
       "5                      0                   1  10:100315722:G:A   \n",
       "...                  ...                 ...               ...   \n",
       "294531             53239                  40    9:22124478:A:G   \n",
       "294532             53239                  40    9:22125504:G:C   \n",
       "294533             53239                  40    9:22125504:G:C   \n",
       "294534             53239                  40    9:22125504:G:C   \n",
       "294535             53239                  40    9:22125504:G:C   \n",
       "\n",
       "                gene_id fpred_max_label  v2g_fpred_max_score  v2g_qtl_score_q  \\\n",
       "0       ENSG00000120054              NA                  NaN              NaN   \n",
       "1       ENSG00000107566              NA                  NaN              NaN   \n",
       "2       ENSG00000107593  intron_variant                  0.1              0.8   \n",
       "4       ENSG00000196072              NA                  NaN              NaN   \n",
       "5       ENSG00000166135              NA                  NaN              NaN   \n",
       "...                 ...             ...                  ...              ...   \n",
       "294531  ENSG00000147883              NA                  NaN              NaN   \n",
       "294532  ENSG00000147889              NA                  NaN              NaN   \n",
       "294533  ENSG00000147883              NA                  NaN              NaN   \n",
       "294534  ENSG00000176399              NA                  NaN              NaN   \n",
       "294535  ENSG00000099810              NA                  NaN              NaN   \n",
       "\n",
       "        v2g_interval_score_q  v2g_distance_score_q  distance_to_tss  \n",
       "0                        NaN                   0.5           233853  \n",
       "1                        NaN                   0.7           129689  \n",
       "2                        NaN                   1.0            14542  \n",
       "4                        NaN                   0.9            29042  \n",
       "5                        0.0                   0.6           213350  \n",
       "...                      ...                   ...              ...  \n",
       "294531                   0.6                   0.8           115173  \n",
       "294532                   0.3                   0.7           130203  \n",
       "294533                   0.6                   0.8           116199  \n",
       "294534                   NaN                   0.4           321320  \n",
       "294535                   0.7                   0.4           322868  \n",
       "\n",
       "[272750 rows x 18 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cred_sets_final=cred_sets_annotated.merge(v2g, how=\"left\", left_on=\"tag_variant_id\", right_on=\"variant_id\")\n",
    "cred_sets_final=cred_sets_final.query('distance_to_tss<=500000', engine='python')\n",
    "cred_sets_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "90b54e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_weighted_mean(df,\n",
    "                          weights = {\n",
    "    'v2g_fpred_max_score': 1,\n",
    "    'v2g_qtl_score_q': 0.66,\n",
    "    'v2g_distance_score_q': 0.33,\n",
    "    'v2g_interval_score_q': 0.33\n",
    "}):\n",
    "    \n",
    "    df.fillna(0, inplace=True)\n",
    "    \n",
    "    # Compute the weighted mean score\n",
    "    weighted_sum = sum(df[col] * weights[col] for col in weights)\n",
    "    total_weight = sum(weights.values())\n",
    "    weighted_mean = weighted_sum / total_weight\n",
    "\n",
    "    # Add the weighted mean score as a new column named 'v2g'\n",
    "    df['v2g'] = weighted_mean\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "88f0aecc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>tag_variant_id</th>\n",
       "      <th>postprob</th>\n",
       "      <th>is95_credset</th>\n",
       "      <th>position</th>\n",
       "      <th>cred_set_min_pos</th>\n",
       "      <th>cred_set_max_pos</th>\n",
       "      <th>cred_set_size_bp</th>\n",
       "      <th>credset_n_variants</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>fpred_max_label</th>\n",
       "      <th>v2g_fpred_max_score</th>\n",
       "      <th>v2g_qtl_score_q</th>\n",
       "      <th>v2g_interval_score_q</th>\n",
       "      <th>v2g_distance_score_q</th>\n",
       "      <th>distance_to_tss</th>\n",
       "      <th>v2g</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000120054</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>233853</td>\n",
       "      <td>0.071121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000107566</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.7</td>\n",
       "      <td>129689</td>\n",
       "      <td>0.099569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000107593</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14542</td>\n",
       "      <td>0.412931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000196072</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>29042</td>\n",
       "      <td>0.128017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCST90002412</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>True</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>100315722</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10:100315722:G:A</td>\n",
       "      <td>ENSG00000166135</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>213350</td>\n",
       "      <td>0.085345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294531</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22124478:A:G</td>\n",
       "      <td>0.005949</td>\n",
       "      <td>True</td>\n",
       "      <td>22124478</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22124478:A:G</td>\n",
       "      <td>ENSG00000147883</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.8</td>\n",
       "      <td>115173</td>\n",
       "      <td>0.199138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294532</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000147889</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3</td>\n",
       "      <td>0.7</td>\n",
       "      <td>130203</td>\n",
       "      <td>0.142241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294533</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000147883</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.8</td>\n",
       "      <td>116199</td>\n",
       "      <td>0.199138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294534</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000176399</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>321320</td>\n",
       "      <td>0.056897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294535</th>\n",
       "      <td>SAIGE_272_11</td>\n",
       "      <td>9:22076072:T:C</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>0.005790</td>\n",
       "      <td>True</td>\n",
       "      <td>22125504</td>\n",
       "      <td>22072265</td>\n",
       "      <td>22125504</td>\n",
       "      <td>53239</td>\n",
       "      <td>40</td>\n",
       "      <td>9:22125504:G:C</td>\n",
       "      <td>ENSG00000099810</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.7</td>\n",
       "      <td>0.4</td>\n",
       "      <td>322868</td>\n",
       "      <td>0.156466</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272750 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            study_id   lead_variant_id    tag_variant_id  postprob  \\\n",
       "0       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "1       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "2       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "4       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "5       GCST90002412  10:100315722:G:A  10:100315722:G:A  0.999993   \n",
       "...              ...               ...               ...       ...   \n",
       "294531  SAIGE_272_11    9:22076072:T:C    9:22124478:A:G  0.005949   \n",
       "294532  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "294533  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "294534  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "294535  SAIGE_272_11    9:22076072:T:C    9:22125504:G:C  0.005790   \n",
       "\n",
       "        is95_credset   position  cred_set_min_pos  cred_set_max_pos  \\\n",
       "0               True  100315722         100315722         100315722   \n",
       "1               True  100315722         100315722         100315722   \n",
       "2               True  100315722         100315722         100315722   \n",
       "4               True  100315722         100315722         100315722   \n",
       "5               True  100315722         100315722         100315722   \n",
       "...              ...        ...               ...               ...   \n",
       "294531          True   22124478          22072265          22125504   \n",
       "294532          True   22125504          22072265          22125504   \n",
       "294533          True   22125504          22072265          22125504   \n",
       "294534          True   22125504          22072265          22125504   \n",
       "294535          True   22125504          22072265          22125504   \n",
       "\n",
       "        cred_set_size_bp  credset_n_variants        variant_id  \\\n",
       "0                      0                   1  10:100315722:G:A   \n",
       "1                      0                   1  10:100315722:G:A   \n",
       "2                      0                   1  10:100315722:G:A   \n",
       "4                      0                   1  10:100315722:G:A   \n",
       "5                      0                   1  10:100315722:G:A   \n",
       "...                  ...                 ...               ...   \n",
       "294531             53239                  40    9:22124478:A:G   \n",
       "294532             53239                  40    9:22125504:G:C   \n",
       "294533             53239                  40    9:22125504:G:C   \n",
       "294534             53239                  40    9:22125504:G:C   \n",
       "294535             53239                  40    9:22125504:G:C   \n",
       "\n",
       "                gene_id fpred_max_label  v2g_fpred_max_score  v2g_qtl_score_q  \\\n",
       "0       ENSG00000120054              NA                  0.0              0.0   \n",
       "1       ENSG00000107566              NA                  0.0              0.0   \n",
       "2       ENSG00000107593  intron_variant                  0.1              0.8   \n",
       "4       ENSG00000196072              NA                  0.0              0.0   \n",
       "5       ENSG00000166135              NA                  0.0              0.0   \n",
       "...                 ...             ...                  ...              ...   \n",
       "294531  ENSG00000147883              NA                  0.0              0.0   \n",
       "294532  ENSG00000147889              NA                  0.0              0.0   \n",
       "294533  ENSG00000147883              NA                  0.0              0.0   \n",
       "294534  ENSG00000176399              NA                  0.0              0.0   \n",
       "294535  ENSG00000099810              NA                  0.0              0.0   \n",
       "\n",
       "        v2g_interval_score_q  v2g_distance_score_q  distance_to_tss       v2g  \n",
       "0                        0.0                   0.5           233853  0.071121  \n",
       "1                        0.0                   0.7           129689  0.099569  \n",
       "2                        0.0                   1.0            14542  0.412931  \n",
       "4                        0.0                   0.9            29042  0.128017  \n",
       "5                        0.0                   0.6           213350  0.085345  \n",
       "...                      ...                   ...              ...       ...  \n",
       "294531                   0.6                   0.8           115173  0.199138  \n",
       "294532                   0.3                   0.7           130203  0.142241  \n",
       "294533                   0.6                   0.8           116199  0.199138  \n",
       "294534                   0.0                   0.4           321320  0.056897  \n",
       "294535                   0.7                   0.4           322868  0.156466  \n",
       "\n",
       "[272750 rows x 19 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_weighted_mean(df=cred_sets_final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7b2f0ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "cred_sets_final.to_csv('OpenTargets_Credible_Sets_Lipids.6May24.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cd887eaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Description of each column\n",
    "column_descriptions = {\n",
    "    'trait_efos': 'EFO Trait/Disase Codes. In this study, only lipid-related codes EFO_0003774, EFO_0004530, EFO_0004611, EFO_0004612, and HP_0003124 are included.',\n",
    "    'trait_reported': 'Reported trait for the GWAS (e.g. Hyperlipidemia)',\n",
    "    'source': 'Source of the study data. Could be SAIGE (UKB) or GCST (GWAS Catalog)',\n",
    "    'study_id': 'ID of the study. There are 39 GWAS catalog studies of lipid traits from the last 10 years + 2 studies from UKB conducted with SAIGE',\n",
    "    'pmid': 'PubMed ID of the study.',\n",
    "    'pub_date': 'Publication date of the study.',\n",
    "    'n_cases': 'Number of cases in the study.',\n",
    "    'num_assoc_loci': 'Number of independent associated loci in the study.',\n",
    "    'lead_variant_id': 'ID of the lead variant for each independent locus.',\n",
    "    'lead_variant_pval': 'P-value of the lead variant in the study.',\n",
    "    'gene_name': 'Gene Name wtihin 500Kb of lead_variant',\n",
    "    'L2G': 'OpenTargets Locus-to-gene score associated with the locus and the gene.',\n",
    "    'coloc_study': 'Colocalization study associated with the variant. Note that colocalization was only attempted if there were overlaps in the credible sets.',\n",
    "    'coloc_type': 'Type of colocalization study (eQTL or pQTL)',\n",
    "    'coloc_tissue': 'Tissue type used in colocalization study.',\n",
    "    'coloc_probe_transcript_id': 'Transcript ID of the probe used in colocalization study.',\n",
    "    'coloc_h0': 'Colocalization hypothesis H0.',\n",
    "    'coloc_h1': 'Colocalization hypothesis H1.',\n",
    "    'coloc_h2': 'Colocalization hypothesis H2.',\n",
    "    'coloc_h3': 'Colocalization hypothesis H3.',\n",
    "    'coloc_h4': 'Colocalization hypothesis H4.',\n",
    "    'coloc_h4_h3': 'Ratio of colocalization hypotheses H4 to H3.',\n",
    "    'coloc_log2_h4_h3': 'Log2 ratio of colocalization hypotheses H4 to H3.',\n",
    "    'postprob': 'Posterior probability of the tag variant being causal.',\n",
    "    'is95_credset': 'Indicator if the tag_variant_id is in the 95% credible set for the locus.',\n",
    "    'cred_set_min_pos': 'Minimum position of variants within the credible set for this locus.',\n",
    "    'cred_set_max_pos': 'Maximum position of variants within the credible set for this locus.',\n",
    "    'cred_set_size_bp': 'Size of the credible set in base pairs.',\n",
    "    'credset_n_variants': 'Number of variants within the credible set for this locus.',\n",
    "    'gene_id' : 'Ensmbl gene ID',\n",
    "    'fpred_max_label': 'Most severe VEP consequence of the tag_variant_id on the gene_id',\n",
    "    'v2g_fpred_max_score' : 'Score based on VEP functional prediction as described at https://github.com/opentargets/v2g_data/blob/master/configs/vep_consequences.tsv',\n",
    "    'v2g_qtl_score_q': 'Maximum quantile of OpenTargets V2G based on -log10(QTL p-value)',\n",
    "    'v2g_interval_score_q': 'Maximum quantile of OpenTargets V2G based on PCHi-C from Javierre et al. (Cell, 2016), Enhancer-TSS correlation from Andersson et al. (Nature, 2014) or DHS-promoter correlation from Thurman et al. (Nature,  2012)',\n",
    "    'v2g_distance_score_q': 'Maximum quantile of OpenTargets V2G based on 1/distance to tss',\n",
    "    'v2g': 'Weighted mean across the v2g quantiles, as described in https://genetics-docs.opentargets.org/our-approach/data-pipeline',\n",
    "    'distance_to_tss' : 'distance to gene transcription start site (TSS) in bp'\n",
    "}\n",
    "\n",
    "# Define credible set related columns\n",
    "credible_set_cols = [\n",
    "    'tag_variant_id', 'postprob',\n",
    "    'is95_credset', 'cred_set_min_pos', 'cred_set_max_pos',\n",
    "    'cred_set_size_bp', 'credset_n_variants',\n",
    "    'gene_id','fpred_max_label', 'v2g_fpred_max_score', 'v2g_qtl_score_q',\n",
    "    'v2g_interval_score_q', 'v2g_distance_score_q', 'v2g', 'distance_to_tss'\n",
    "]\n",
    "\n",
    "# Define other columns\n",
    "other_cols = set(column_descriptions.keys()) - set(credible_set_cols)\n",
    "\n",
    "# Write descriptions to README.txt\n",
    "with open('README.txt', 'w') as readme_file:\n",
    "  readme_file.write(\"\\n# These two files can be joined on study_id and lead_variant_id \\n\\n\")\n",
    "  readme_file.write(\"# Credible Set File Columns\\n\\n\")\n",
    "  for col in credible_set_cols:\n",
    "    readme_file.write(f\"{col}: {column_descriptions.get(col, 'Description not available.')}\\n\")\n",
    "\n",
    "  readme_file.write(\"\\n# Other File Columns\\n\\n\")\n",
    "  for col in sorted(other_cols):\n",
    "    readme_file.write(f\"{col}: {column_descriptions.get(col, 'Description not available.')}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cell_programs",
   "language": "python",
   "name": "cell_programs"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
